### PR TITLE
add java and python solutions to quantum superposition

### PR DESCRIPTION
**Pull request by 876pol**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

### CODE REVIEW

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 1]
```diff
@@ -91,3 +93,147 @@ int main() {
...
+public class QuantumSuperposition {
+	static int n[] = new int[2];
+	static int m[] = new int[2];
+	static ArrayList<Integer> g[][] = new ArrayList[2][1001];
+	static ArrayList<Integer> back[][] = new ArrayList[2][1001];
+	static boolean dp[][][] = new boolean[2][1001][2001];
+	static Scanner scanner = new Scanner(System.in);
+
+	static void gen(int x) {
+		int[] in_degree = new int[1001];
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:114]: java use camelCase + magic number

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 2]
```diff
@@ -91,3 +93,147 @@ int main() {
...
+			int a = scanner.nextInt();
+			int b = scanner.nextInt();
+			g[x][a].add(b);
+			back[x][b].add(a);
+			in_degree[b]++;
+		}
+		// finding length of routes of first universe
+		Queue<Integer> q = new LinkedList<>();
+		for (int i = 0; i <= n[x]; i++) {
+			if (in_degree[i] == 0) q.add(i);
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:125]: braces

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 3]
```diff
@@ -91,3 +93,147 @@ int main() {
...
+</CPPSection>
+
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class QuantumSuperposition {
+	static int n[] = new int[2];
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:106]: use java-style array decls

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 4]
```diff
@@ -91,3 +93,147 @@ int main() {
...
+				if (in_degree[next] == 0) q.add(next);
+			}
+		}
+	}
+
+	public static void main(String[] args) throws IOException {
+		n[0] = scanner.nextInt();
+		n[1] = scanner.nextInt();
+		m[0] = scanner.nextInt();
+		m[1] = scanner.nextInt();
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:147]: think the convention is to use bufferedreader?

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 5]
```diff
@@ -91,3 +93,147 @@ int main() {
...
+
+public class QuantumSuperposition {
+	static int n[] = new int[2];
+	static int m[] = new int[2];
+	static ArrayList<Integer> g[][] = new ArrayList[2][1001];
+	static ArrayList<Integer> back[][] = new ArrayList[2][1001];
+	static boolean dp[][][] = new boolean[2][1001][2001];
+	static Scanner scanner = new Scanner(System.in);
+
+	static void gen(int x) {
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:113]: quick doc comment for this func

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 6]
```diff
@@ -91,3 +93,147 @@ int main() {
...
+<PySection>
+
+```py
+from collections import deque
+
+n = [0, 0]
+m = [0, 0]
+g = [[[] for _ in range(1001)] for _ in range(2)]
+back = [[[] for _ in range(1001)] for _ in range(2)]
+dp = [[[False for _ in range(2001)] for _ in range(1001)] for _ in range(2)]
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:189]: magic #

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 7]
```diff
@@ -91,3 +93,147 @@ int main() {
...
+
+
+def gen(x):
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:192]: put doc comment here as well

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 8]
```diff
@@ -91,3 +93,147 @@ int main() {
...
+			if dp[1][n[1]][j]:
+				ans[i + j] = True
+
+Q = int(input())
+for _ in range(Q):
+	a = int(input())
+	if ans[a]:
+		print("Yes")
+	else:
+		print("No")
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:236]:
```suggestion
	print("Yes" if ans[a] else "No")
```

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 9]
```diff
@@ -91,3 +93,147 @@ int main() {
...
+			}
+		}
+		int Q = scanner.nextInt();
+		for (int i = 0; i < Q; i++) {
+			int a = scanner.nextInt();
+			if (ans[a]) {
+				System.out.println("Yes");
+			} else {
+				System.out.println("No");
+			}
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:172]:
```suggestion
			System.out.println(ans[a] ? "Yes" : "No");
```

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 10]
```diff
@@ -91,3 +93,147 @@ int main() {
...
+		// preprocessing all possible sums between the universes
+		boolean[] ans = new boolean[2001];
+		for (int i = 0; i < 1001; i++) {
+			if (dp[0][n[0]][i]) {
+				for (int j = 0; j < 1001; j++) {
+					if (dp[1][n[1]][j]) ans[i + j] = true;
+				}
+			}
+		}
+		int Q = scanner.nextInt();
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:165]: make lowercase

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 11]
```diff
@@ -28,66 +28,238 @@ Once we know all possible path lengths for each universe, we can find all
...
+
 ```cpp
 #include <bitset>
 #include <iostream>
 #include <queue>
 #include <vector>
 using namespace std;
 
+static const int MAX_N = 1001;
+static const int MAX_Q = 2001;
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:42]: why's the +1 necessary

Comment (876pol) [reply solutions/gold/kattis-quantumsuperposition.mdx:42]: the graph is 1-indexed, and the dp needs states from [0, N] and [0, Q]

Comment (SansPapyrus683) [reply solutions/gold/kattis-quantumsuperposition.mdx:42]: can you make the graph 0-indexed
also use `MAX_N + 1` and `MAX_Q + 1` if you don't wanna

Comment (876pol) [reply solutions/gold/kattis-quantumsuperposition.mdx]: alr i did the +1 thing

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 12]
```diff
@@ -28,66 +28,238 @@ Once we know all possible path lengths for each universe, we can find all
...
-vector<int> g[2][1001];
-vector<int> back[2][1001];
-bitset<2001> dp[2][1001];
+vector<int> g[2][MAX_N + 1];
+vector<int> back[2][MAX_N + 1];
+bitset<MAX_Q + 1> dp[2][MAX_N + 1];
 
+/*
+ * Generates the topological sort, filling in the DP array in the process.
+ */
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:51]:
```suggestion
/** Generates the topological sort, filling in the DP array in the process. */
```

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 13]
```diff
@@ -28,66 +28,238 @@ Once we know all possible path lengths for each universe, we can find all
...
+
+		// `node` can be visited in `i + 1` steps if `before` can in `i` steps.
+		if (back[x][node].empty()) { dp[x][node][0] = 1; }
+		for (int before : back[x][node]) { dp[x][node] |= dp[x][before] << 1; }
 	}
 }
 
 int main() {
 	ios_base::sync_with_stdio(0);
 	cin.tie(0);
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:82]: remove if runs w/o

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 14]
```diff
@@ -28,66 +28,238 @@ Once we know all possible path lengths for each universe, we can find all
...
+
+	// Preprocess for all queries: ans[x] is true if the sum of steps
+	// in both universes can be equal to x.
+	bitset<MAX_Q + 1> ans;
+	for (int i = 0; i <= MAX_N; i++) {
+		if (dp[0][n[0]][i]) { ans |= dp[1][n[1]] << i; }
+	}
+
 	int Q;
 	cin >> Q;
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:96]: make lowercase

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/gold/kattis-quantumsuperposition.mdx (hunk 13 above)
Comment (876pol) [reply solutions/gold/kattis-quantumsuperposition.mdx:80]: remove the entire lines? it doesnt work if i do

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 15]
```diff
@@ -28,66 +28,236 @@ Once we know all possible path lengths for each universe, we can find all
...
+from collections import deque
+
+n = [0, 0]
+m = [0, 0]
+g = [[[] for _ in range(MAX_N + 1)] for _ in range(2)]
+back = [[[] for _ in range(MAX_N + 1)] for _ in range(2)]
+dp = [[[False for _ in range(MAX_Q + 1)] for _ in range(MAX_N + 1)] for _ in range(2)]
+
+# Generates the topological sort, filling in the DP array in the process.
+def gen(x):
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:213]:
```suggestion
def gen(x):
"""Generates the topological sort, filling in the DP array in the process."""
```

File: solutions/gold/kattis-quantumsuperposition.mdx [hunk 16]
```diff
@@ -28,66 +28,236 @@ Once we know all possible path lengths for each universe, we can find all
...
+# in both universes can be equal to x.
+ans = [False for _ in range(MAX_Q + 1)]
+for i in range(MAX_N + 1):
+	if dp[0][n[0]][i]:
+		for j in range(MAX_N + 1):
+			if dp[1][n[1]][j]:
+				ans[i + j] = True
+
+q = int(input())
+for _ in range(q):
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-quantumsuperposition.mdx:257]:
```suggestion
for _ in range(int(input())):
```

**Review by SansPapyrus683 — APPROVED** [review]

File: solutions/gold/kattis-quantumsuperposition.mdx (hunk 13 above)
Comment (ryanchou-dev) [reply solutions/gold/kattis-quantumsuperposition.mdx:80]: he means the fast i/o fyii

**Review by ryanchou-dev — APPROVED** [review]

**Pull request merged**